pyomo/contrib/multisolve/multisolve.py

```diff
@@ -0,0 +1,146 @@
+from __future__ import division
+
+import logging
+import time
+import timeit
+
+_pathos_available = True
+try:
+    # Pathos is required because a dill-friendly multiprocessing alternative was necessary.
+    from pathos.pools import _ProcessPool as Pool
+except ImportError:
+    _pathos_available = False
+
+from pyomo.contrib.gdpopt.data_class import GDPoptSolveData
+from pyomo.contrib.gdpopt.util import create_utility_block, copy_var_list_values
+from pyomo.core import Suffix
+from pyomo.common.config import (
+    ConfigBlock, ConfigValue, add_docstring_list
+)
+from pyomo.opt import SolverFactory, SolverResults
+from pyomo.opt import TerminationCondition as tc
+
+logger = logging.getLogger('pyomo.contrib.multistart')
+
+
+@SolverFactory.register('multisolve',
+                        doc='Naive Parallel Solver Harness for Pyomo')
+class Multisolve(object):
+    """Solver wrapper that spawns multiple parallel attempts at solution.
+
+    CAUTION: instability is possible because parallel processing gets tricky.
+
+    Keyword arguments below are specified for the ``solve`` function.
+
+    """
+
+    CONFIG = ConfigBlock()
+    CONFIG.declare("solvers", ConfigValue(
+        default=['ipopt', ], domain=list,
+        description="Specify the list of solvers to use.",
+    ))
+    CONFIG.declare("solver_args", ConfigValue(
+        default=[{}, ],
+        description="List of dictionary of keyword arguments to pass to the solver."
+    ))
+    CONFIG.declare("time_limit", ConfigValue(
+        default=float('inf'),
+        description="Time limit on execution (seconds)"
+    ))
+    CONFIG.declare("integer_tolerance", ConfigValue(
+        default=1e-4,
+        description="Tolerance on integrality"
+    ))
+
+    __doc__ = add_docstring_list(__doc__, CONFIG)
+
+    def available(self, exception_flag=True):
+        """Check if solver is available.
+
+        TODO: Sub-solvers may not always be available, and so this should reflect that possibility.
+
+        """
+        return _pathos_available
+
+    def solve(self, model, **kwds):
+        config = self.CONFIG(kwds.pop('options', {}))
+        config.set_value(kwds)
+
+        if not _pathos_available:
+            raise ImportError("Multisolve requires the `pathos` package.")
+
+        solve_data = GDPoptSolveData()
+        with create_utility_block(model, 'multisolve_utils', solve_data):
+            results = []
+
+            with Pool(processes=2, maxtasksperchild=1) as pool:
```

Comment:
It looks like the use of a multiprocessing pool as a context manager is only supported in Python 3.  I am not sure why you need to use it in a `with` clause, either.  It seems if you just replace this with
```
pool = Pool(processes=2, maxtasksperchild=1)
```
then things work in Python 2.7 and 3.x.

Reply:
Makes sense. I think the context manager might automatically `terminate` the pool at the end? I can't remember exactly.

Reply:
Fair enough.  If there is something that happens on `__exit__`, then we should be able to implement it in a portable way with an explicit:
```python
try:
    # use the pool
finally:
   # terminate the pool
```